Lib/test/test_email/test_email.py

```diff
@@ -728,6 +728,46 @@ def test_nonascii_add_header_with_tspecial(self):
             "attachment; filename*=utf-8''Fu%C3%9Fballer%20%5Bfilename%5D.ppt",
             msg['Content-Disposition'])
 
+    def test_invalid_header_names(self):
+        invalid_headers = [
+            ('Invalid Header', 'contains space'),
+            ('Tab\tHeader', 'contains tab'),
+            ('Colon:Header', 'contains colon'),
+            ('', 'Empty name'),
+            (' LeadingSpace', 'starts with space'),
+            ('TrailingSpace ', 'ends with space'),
+        ]
+        for name, value in invalid_headers:
+            with self.subTest(name=name, problem=value):
+                with self.assertRaises(ValueError) as cm:
+                    Message().add_header(name, value)
+                self.assertIn(f"Invalid header field name {name!r}", str(cm.exception))
```

Comment:
It is better not to so highly constrain the error message.  We really don't care what exactly the error message says.  The important bits are the words 'invalid', 'header', and 'name', and the repr of the header value.  To make the test place only those constraints on the error message, we can write:
```
msg = str(cm.exception)
self.assertRegex(
    msg,
    '(?i)(?=.*invalid)(?=.*header)(?=.*name)',
    )
self.assertIn(f"{name!r}", msg)
````
This uses lookahead assertions to check for each of the substrings in any position in the target string.  Using this as the tests, both versions of your error message pass, as would several other variations that would be equally valid.  And that makes it easy to see that if you delete your first 'if' test in the validator, all the tests still pass.